### DISCUSSION TITLE
Question about Clamping of Neurons During Supervised Learning

### DISCUSSION THREAD

**Discussion by DS-567** [BindsNET/bindsnet#600] [General]
I have a question about supervised learning in BindsNet using the clamp and STDP based learning rules.

So my understanding of how BindsNet implements STDP (PostPre) in supervised learning is by forcing the target neuron for that class to spike continuously for every simulation timestep, which is done with the clamp parameter in the network.run function. The reason for this is because STDP needs output spikes present to determine pre and post synaptic spike times to update weights at the synapses. 

But for the project I am working on, I am presenting a input sample to the network which only consists of two spikes. I only want the target neuron to spike once as soon as both input spikes have occurred, so the target neuron is basically trained to recognise the two spikes from only two inputs by producing one output spike. For example, if the first input spike occurs at 0ms, and the second input spike at 20ms, I would like the target neuron to spike at 21ms. So I want to clamp the target neuron to spike only at 21ms and not every millisecond between 0 and 21 if that makes sense. And I then train all other target neurons for all the patterns that I want to use. 

**_Is there a way I can only clamp the target neuron to spike at the same time as the second input spike comes in, in order for the weights to increase with STDP? Instead of having it clamp every timestep of the input sample._**

Currently I am repeating with multiple epochs and carefully monitoring the target neurons output spike and stopping as soon as it spikes once after the second input spike happens. But both weights are increasing at the same rate.

**Comment by Hananel-Hazan** [comment]
Your description for the use for the STDP to modify synaptic weights is fairly accurate and you can find similar example here [Supervised MNIST](https://github.com/BindsNET/bindsnet/blob/master/examples/mnist/supervised_mnist.py). However that is not the only possible usage. Please check out the [MSTDP](https://github.com/BindsNET/bindsnet/blob/2104e82edd1b58dde6631aa33f822759536d773d/bindsnet/learning/learning.py#L1428) reward rule that seems more suitable to your scenario. Using the MSTDP you can devise a reward signal that activate only when the two neurons are spiking at the same time.
